### ISSUE TITLE
discussion - self-completing commends

### ISSUE THREAD

**Issue by akdor1154** [jdx/usage#5]
heya,

this looks fab! One thing I've been wondering about is some sort of self-documenting/completing mechanism - e.g. some way for bins/scripts to act as their own completer.

having a way to get a usage spec out of a script/bin (e.g. your double-shebang mechanism) is one half, but the other is for the command itself to be recursively involved in completions. Think e.g. a field that wants an s3:// url, or an scp remotehost:/remotepath field. Not fleshed out sorry, I'll come back to this :)

**Comment by jdx** [comment]
yes, this is table-stakes. In fact what I need usage for currently is the ability to "mount" one CLI into another for `mise run ...`. I haven't fully figured out how this will work but right now but in usage-cli I'm thinking the commands to build completions and stuff could accept a command string, e.g.:

```sh
$ usage complete-word -x "mycli --usage"
```

And it would exec `mycli --usage`, get the dynamic definition, and be able to complete against it.

But it may make sense to configure this by putting it into the spec itself, maybe we could add a param like this:

```sh
fetch_spec "mycli --usage"
```

I don't love that name but something like that.

**Comment by Wikiwix** [comment]
I would be very interested in this feature :)
Is this still planned?

**Comment by jdx** [comment]
Yes, this is implemented now:

- Dynamic command specs are supported with [`mount run="..."`](https://usage.jdx.dev/spec/reference/cmd.html#mounting-dynamic-commands).
- Dynamic argument values are supported with [`complete ... run="..."`](https://usage.jdx.dev/spec/reference/complete.html).
- mise mounts task specs into `mise run` completion; this landed in [jdx/mise#2661](https://github.com/jdx/mise/pull/2661).

Closing as completed.

*AI-assisted — Tool: Codex; model: openai/gpt-5; version: unavailable.*
